src/os/bluestore/Allocator.h

```diff
@@ -96,27 +295,30 @@ class Allocator {
   // - amount of extents aligned with allocation boundary
   // - amount of allocation units in aligned extents
   //
-  struct free_state_hist_bucket {
-    static const size_t base_bits = 12;
-    static const size_t base = 1ull << base_bits;
-    static const size_t mux = 2;
+  class FreeStateHistogram {
+    const Allocator::ExtentCollectionTraits myTraits;
+    enum {
+      BASE_BITS = 12, // 4096 bytes
+      FACTOR = 2,
+    };
+    struct free_state_hist_bucket {
+      size_t total = 0;
+      size_t aligned = 0;
+      size_t alloc_units = 0;
+    };
+    std::vector<free_state_hist_bucket> buckets;
+  public:
 
-    size_t total = 0;
-    size_t aligned = 0;
-    size_t alloc_units = 0;
+    FreeStateHistogram(size_t num_buckets)
+      : myTraits(num_buckets, BASE_BITS, FACTOR) {
+      buckets.resize(num_buckets);
+    }
 
-    // returns upper bound of the bucket
-    static size_t get_max(size_t bucket, size_t num_buckets) {
-      return
-        bucket < num_buckets - 1 ?
-          base << (mux * bucket) :
-          std::numeric_limits<uint64_t>::max();
-    };
+    void record_extent(uint64_t alloc_unit, uint64_t off, uint64_t len);
+    void foreach(
```

Comment:
`foreach()` nicely isolates user from implementation.
But if we do such a thing, at least rudimentary information in declaration should give some light on meaning of
`uint64_t, uint64_t, uint64_t, uint64_t`.

Reply:
done